### PR TITLE
Bracket model introduction

### PR DESCRIPTION
**Pull request by josef-vlach**
Most important in this PRs is change in `FormModel` where
```
pagesWithIndex: List[(PageModel[A], SectionNumber)],
```
has been superseeded by:
```
brackets: BracketsWithSectionNumber[A],
```

`BracketsWithSectionNumber` holds data about where pages in the form comes from, ie. from NonRepeating section, Repeating section or from Add To List.

### CODE REVIEW

File: app/uk/gov/hmrc/gform/models/Bracket.scala [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    def firstSectionNumber: SectionNumber = singletons.head.sectionNumber
+
+  }
+
+  case class NonRepeatingPage[A <: PageMode](
+    singleton: Singleton[A],
+    sectionNumber: SectionNumber,
+    source: Section.NonRepeatingPage
+  ) extends Bracket[A] {
+    def map[B <: PageMode](f: Singleton[A] => Singleton[B], g: Repeater[A] => Repeater[B]): NonRepeatingPage[B] =
```
Comment (girishkamat) [inline app/uk/gov/hmrc/gform/models/Bracket.scala:105]: `g: Repeater[A] => Repeater[B]` seems unused. is it required?

File: app/uk/gov/hmrc/gform/models/Bracket.scala [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    def toPlainBracket: BracketPlain[A] = BracketPlain.NonRepeatingPage(singleton, source)
+
+    def hasSectionNumber(sectionNumber: SectionNumber): Boolean = this.sectionNumber === sectionNumber
+  }
+
+  case class RepeatingPage[A <: PageMode](
+    singletons: NonEmptyList[SingletonWithNumber[A]], // There must be at at least one page in repeating page bracket
+    source: Section.RepeatingPage
+  ) extends Bracket[A] {
+    def map[B <: PageMode](f: Singleton[A] => Singleton[B], g: Repeater[A] => Repeater[B]): RepeatingPage[B] =
```
Comment (girishkamat) [inline app/uk/gov/hmrc/gform/models/Bracket.scala:121]: `g: Repeater[A] => Repeater[B]` seems unused

File: app/uk/gov/hmrc/gform/models/Bracket.scala (hunk 1 above)
Comment (josef-vlach) [reply app/uk/gov/hmrc/gform/models/Bracket.scala:105]: `map` is `override` method.

File: app/uk/gov/hmrc/gform/models/Bracket.scala (hunk 2 above)
Comment (josef-vlach) [reply app/uk/gov/hmrc/gform/models/Bracket.scala:121]: `map` is `override` method.

**Review by girishkamat — APPROVED** [review]

**Pull request merged**